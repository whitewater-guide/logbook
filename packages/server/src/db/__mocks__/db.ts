import {
  DatabasePoolConnectionType,
  DatabasePoolType,
  createPool,
} from 'slonik';

import { Client } from 'pg';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: `logbook_test_${process.env.JEST_WORKER_ID}`,
  password: process.env.POSTGRES_PASSWORD,
};

let _pool: DatabasePoolType | undefined;

export const db: () => DatabasePoolConnectionType = () => {
  if (!_pool) {
    throw new Error('no test pool found');
  }
  return _pool;
};

export const setupDB = async () => {
  const workerID = process.env.JEST_WORKER_ID;
  if (!workerID) {
    throw new Error('JEST_WORKER_ID not found');
  }
  const client = new Client({
    host: process.env.POSTGRES_HOST,
    user: process.env.POSTGRES_USER,
    database: process.env.POSTGRES_DB,
    password: process.env.POSTGRES_PASSWORD,
  });
  try {
    await client.connect();
    await client.query(`DROP DATABASE IF EXISTS logbook_test_${workerID}`);
    await client.query(
      `CREATE DATABASE logbook_test_${workerID} TEMPLATE logbook_test_template`,
    );
  } finally {
    await client.end();
  }

  _pool = createPool(
    `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  );
};

export const teardownDB = async () => {
  await _pool?.end();
};
