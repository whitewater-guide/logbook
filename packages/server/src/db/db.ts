import { createPool, DatabasePoolConnectionType } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
