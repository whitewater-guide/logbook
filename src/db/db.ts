import {
  DatabasePoolConnectionType,
  createPool,
  createTypeParserPreset,
} from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database: process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

const pool = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);

export const db: () => DatabasePoolConnectionType = () => pool;
export const setupDB = () => Promise.resolve();
export const teardownDB = async () => {
  await pool.end();
};
