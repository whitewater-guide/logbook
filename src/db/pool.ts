import { createPool, createTypeParserPreset } from 'slonik';

const cfg = {
  host: process.env.POSTGRES_HOST,
  user: process.env.POSTGRES_USER,
  database:
    process.env.NODE_ENV === 'test'
      ? `logbook_test_${process.env.JEST_WORKER_ID}`
      : process.env.POSTGRES_DB,
  password: process.env.POSTGRES_PASSWORD,
};

const typeParsers = createTypeParserPreset().filter(
  ({ name }) => !['name', 'timestamp', 'timestamptz'].includes(name),
);

export const db = createPool(
  `postgresql://${cfg.user}:${cfg.password}@${cfg.host}/${cfg.database}`,
  {
    typeParsers,
  },
);
