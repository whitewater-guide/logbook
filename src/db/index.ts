import { db } from './pool';

export default db;
