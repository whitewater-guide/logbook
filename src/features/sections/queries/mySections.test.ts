import {
  ListMySectionsQuery,
  ListMySectionsQueryVariables,
} from './mySections.test.generated';
import { USER_1, USER_2 } from '~/test/fixtures';
import { setupDB, teardownDB } from '~/db';

import { gql } from 'apollo-server';
import { runQuery } from '~/test/apollo-helpers';

beforeEach(setupDB);
afterEach(teardownDB);

const query = gql`
  query listMySections($filter: SectionsFilter, $page: Page) {
    mySections(filter: $filter, page: $page) {
      edges {
        node {
          id

          region
          river
          section
          difficulty

          putIn {
            lat
            lng
          }
          takeOut {
            lat
            lng
          }

          upstreamId
          upstreamData

          createdAt
          updatedAt
        }
        cursor
      }
      pageInfo {
        endCursor
        hasMore
      }
    }
  }
`;

it('should match snapshot', async () => {
  const result = await runQuery<
    ListMySectionsQuery,
    ListMySectionsQueryVariables
  >(query, {}, USER_1);
  expect(result).toMatchSnapshot();
});

it.each([
  ['anon', undefined],
  ['other user', USER_2],
])('% should fail', async () => {
  const result = await runQuery<
    ListMySectionsQuery,
    ListMySectionsQueryVariables
  >(query);
  expect(result.errors).toBeTruthy();
  expect(result.data?.mySections).toBeNull();
});
