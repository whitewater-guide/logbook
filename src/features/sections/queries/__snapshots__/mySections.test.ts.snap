// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
Object {
  "data": Object {
    "mySections": Object {
      "edges": Array [
        Object {
          "cursor": "djE6Mw==",
          "node": Object {
            "createdAt": 2020-01-03T00:00:00.000Z,
            "difficulty": 2.5,
            "id": "7565d9ee-0fc0-4b49-be34-a62d59223067",
            "putIn": null,
            "region": "Norway",
            "river": "Sjoa",
            "section": "Asengjuvet",
            "takeOut": null,
            "updatedAt": 2020-01-03T00:00:00.000Z,
            "upstreamData": null,
            "upstreamId": null,
          },
        },
        Object {
          "cursor": "djE6Mg==",
          "node": Object {
            "createdAt": 2020-01-02T00:00:00.000Z,
            "difficulty": 3,
            "id": "e7d03e01-4927-4ae1-82bf-a9e195e59db6",
            "putIn": null,
            "region": "Norway",
            "river": "Sjoa",
            "section": "Playrun",
            "takeOut": null,
            "updatedAt": 2020-01-02T00:00:00.000Z,
            "upstreamData": null,
            "upstreamId": null,
          },
        },
        Object {
          "cursor": "djE6MQ==",
          "node": Object {
            "createdAt": 2020-01-01T00:00:00.000Z,
            "difficulty": 3.5,
            "id": "45e70501-5d20-449f-b854-52b06b138831",
            "putIn": Object {
              "lat": 49,
              "lng": 0,
            },
            "region": "Norway",
            "river": "Sjoa",
            "section": "Amot",
            "takeOut": Object {
              "lat": 49,
              "lng": 1,
            },
            "updatedAt": 2020-01-01T00:00:00.000Z,
            "upstreamData": Object {
              "upstream": "upstream section 1 data",
            },
            "upstreamId": "d749dbc1-166f-41d1-91ac-e4658d30d90f",
          },
        },
      ],
      "pageInfo": Object {
        "endCursor": "djE6MQ==",
        "hasMore": false,
      },
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
